Extend OTP support to also allow using FTM push

fvanderveen opened the pull request:
In our organization, the fortinet server is configured to be able to send a push notification to the FortiToken app. I figured I could give it a shot to add this feature to this client :).

Checking the requests the official client sends for this, I found it sends `ftmpush=1` instead of the `code=<OTP>&code2=` data in the `logincheck` post.

I've done my best to keep the code around this as clean as I could, but my C++ isn't that strong.  
The code does what I want it to do though; it sends me a push notification if I don't pre-configure the OTP, and uses the given OTP if I run `openfortivpn` with the `--otp=<OTP>` option.

DimitriPapadopoulos commented:
As I'm new to this, some context:
* [Technical Tip: FortiToken Push on FortiAuthenticator - operation flow and details](https://kb.fortinet.com/kb/documentLink.do?externalID=FD45559)
* [SSL VPN with FortiToken mobile push authentication](https://docs.fortinet.com/document/fortigate/6.2.4/cookbook/183204/ssl-vpn-with-fortitoken-mobile-push-authentication)
* [Configuring FortiToken Mobile](https://docs.fortinet.com/document/fortigate/6.2.4/cookbook/776309/configuring-fortitoken-mobile)

mrbaseman commented:
Looks good to me.

mrbaseman commented:
Maybe the help output and the man page should be updated?

fvanderveen commented on src/http.c:
```diff
@@ -597,7 +597,7 @@ int auth_log_in(struct tunnel *tunnel)
 	char polid[32] = { '\0' };
 	char group[128] = { '\0' };
-	char data[1024], token[128], tokenresponse[256];
+	char data[1024], token[128], tokenresponse[256], tokenparams[320];
```
I tried that; but I got a compiler warning when trying to fill `tokenresponse` on line 696:
```
snprintf(tokenparams, sizeof(tokenparams),
	         "code=%s&code2=",
	         tokenresponse);
```
If I recall correctly; the compiler emitted a warning that possibly a too-large amount of data would be put in...

But I'm not that well-versed in C++, so it might very well be possible :).

DimitriPapadopoulos replied:
Perhaps I can try to fix this for you - just increase the size of `tokenresponse` from 256 to 320. I shall modify your branch directly, it's usually possible in the context of a pull request. WIll you then review my changes and check they still work?

fvanderveen replied:
If I understood the warning correctly; doing the following will always complain:
```
			url_encode(tokenresponse, cfg->otp);
			snprintf(tokenresponse, sizeof(tokenresponse),
			         "code=%s&code2=",
			         tokenresponse);
```

The compiler complains with:
```
src/http.c:696:13: warning: passing argument 1 to restrict-qualified parameter aliases with argument 4 [-Wrestrict]
  696 |    snprintf(tokenresponse, sizeof(tokenresponse),
      |             ^~~~~~~~~~~~~
  697 |             "code=%s&code2=",
  698 |             tokenresponse);
      |             ~~~~~~~~~~~~~
src/http.c:697:19: warning: ‘%s’ directive output may be truncated writing up to 255 bytes into a region of size 251 [-Wformat-truncation=]
  697 |             "code=%s&code2=",
      |                   ^~
  698 |             tokenresponse);
      |             ~~~~~~~~~~~~~
```

This is why I added the third field, as I had no idea how to otherwise solve this :(. Increasing the size of `tokenresponse` will only change the truncation bytes.

DimitriPapadopoulos replied:
Ah right, it looks like we do need two variables anyway.

Then there's just the documentation issue only, new options need to be documented in `src/main.c` and `doc/openfortivpn.1.in`.

fvanderveen commented:
> Maybe the help output and the man page should be updated?

@mrbaseman @DimitriPapadopoulos I didn't really add any options. Though this comment made me think it might be useful to add an opt-out option for this. So I added a commit that adds `--no-ftm-push` to disable using this 2fa mechanism.

I tried figuring out if I could somehow add this to the config file as well; but I couldn't find a nice way to do so. If either of you think it's necessary to be possible through the config file as well; some help on what direction would be nice.

mrbaseman commented:
the config file is loaded [here in load_config](https://github.com/adrienverge/openfortivpn/blob/41bbbb9651bcb8fdc977c6a6756964838c1c680d/src/config.c#L207-L433) in config.c and merged with the command line options at a later stage

fvanderveen commented:
@mrbaseman thanks; I think I got everything now. Please have a look again :).

DimitriPapadopoulos commented on src/http.c:
```diff
@@ -658,20 +658,37 @@ int auth_log_in(struct tunnel *tunnel)
+			         "code=%s&code2=",
+			         tokenresponse);
 		}
 
```
Just want to make sure, what should happen if the server sends "ftpm_push" but `(cfg->otp[0] != '\0')`?

fvanderveen replied:
If the config or command line specifies the OTP, I would expect that to have precedence. This is how I coded it (and manually tested it).

My reasoning; if you decided to use OTP explicitly, it doesn't seem to make sense to still use the FTM push, as both mechanisms are separate. (The forti vpn client gives you a GUI with the OTP field + submit button and a separate button for FTM push, when I tried it against my company's VPN.)

fvanderveen commented:
I rebased this due to a merge conflict in `http.c`. It came from the hostcheck support that was recently merged.

Did a small smoke test for FTM push, as well as checking if `--otp=<otp>` and `--no-ftm-push` still worked as expected.

One thing I noticed, while testing after the rebase, is that our fortinet server returned a http/405 when sending both `ftmpush=1` and `magic=<value>` in the same logincheck request.  
Since it was happy to eat it when sending `code=<otp>&code2=`, I moved this parameter to that part of the code; as I suspect this check functionality is only used (or supported) when sending an OTP token back for 2FA.

DimitriPapadopoulos commented:
Looks good to me. Thank you for taking the extra time to re-check after the last rebase. Would you agree to squashing these commits into a single commit?

Otherwise it might be worth intercepting communication between FortiClient and the Fortinet appliance to make sure there are no other parameters like `magic=` that are omited with `ftmpush=1`. I'll have a look if I can get hold of a recent FortiClient at some point and check relevant traffic.

Finally we still have this issue with Travis CI. I really don't know how to fix it without looking into the repository settings. Perhaps this branch has inherited incorrect ties to Travis CI from before #732.

fvanderveen commented:
@DimitriPapadopoulos looked at what the windows forti vpn client does; and even though it receives a `portal` and `magic` parameter; neither is actually sent back for the `ftmpush=1` request.

I don't want to assume that this is always going to be the case. It could be something simple as the client recognizing `portal=maz-tunnel` as the default, and simply not sending it back for that reason.

As I don't have access to configure this server, I cannot really test what the multi-portal functionality would do with ftmpush. (I'd like to err on the side of caution here; and just send it along. It doesn't seem to break anything here.)


I'm fine with squashing these commits, don't think there is anything too useful to keep in the separate commit messages that isn't also mentioned in this PR.

DimitriPapadopoulos commented:
Great. Just squash the commits then, I think we're ready to integrate them. Thank you so much for your patience.

fvanderveen commented:
@DimitriPapadopoulos ah, I thought it would be an option in github when merging the branch; maybe I'm spoiled by this option existing in gitlab :).

I've squashed the commits, and added some description into the commit message, just in case that helps someone in the future.

mrbaseman commented:
@DimitriPapadopoulos Squashing the commits during merge is also possible here. Unfortunately Github is now waiting for Status to be reported back, I think from Travis ci

DimitriPapadopoulos commented:
Oh sure, it's an option in GitHub too. But I think we cannot both rebase and squash - and sometimes you need both by the time you merge.

Yes, Travis CI still doesn't work on this PR. Perhaps we can test this commit locally and then ask @adrienverge to merge.

adrienverge commented:
> But I think we cannot both rebase and squash

I believe the "Squash and merge" button actually *squashes and rebases* (so it's awkwardly named).

> Yes, Travis CI still doesn't work on this PR. Perhaps we can test this commit locally and then ask @adrienverge to merge.

Sure, just ping me when needed.

DimitriPapadopoulos commented:
@adrienverge The `otp/add-ftm-push-support` branch passes tests. Can you please merge?

adrienverge commented:
Sure!
